data/beaconrestapi/src/main/java/tech/pegasys/teku/beaconrestapi/handlers/v1/beacon/GetBlockHeaders.java

```diff
@@ -0,0 +1,116 @@
+import io.javalin.http.Context;
+import io.javalin.http.Handler;
+import io.javalin.plugin.openapi.annotations.HttpMethod;
+import io.javalin.plugin.openapi.annotations.OpenApi;
+import io.javalin.plugin.openapi.annotations.OpenApiContent;
+import io.javalin.plugin.openapi.annotations.OpenApiParam;
+import io.javalin.plugin.openapi.annotations.OpenApiResponse;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import org.apache.tuweni.bytes.Bytes32;
+import org.jetbrains.annotations.NotNull;
+import tech.pegasys.teku.api.ChainDataProvider;
+import tech.pegasys.teku.api.DataProvider;
+import tech.pegasys.teku.api.response.v1.beacon.GetBlockHeadersResponse;
+import tech.pegasys.teku.beaconrestapi.SingleQueryParameterUtils;
+import tech.pegasys.teku.beaconrestapi.handlers.AbstractHandler;
+import tech.pegasys.teku.beaconrestapi.schema.BadRequest;
+import tech.pegasys.teku.infrastructure.async.SafeFuture;
+import tech.pegasys.teku.infrastructure.unsigned.UInt64;
+import tech.pegasys.teku.provider.JsonProvider;
+
+public class GetBlockHeaders extends AbstractHandler implements Handler {
+  public static final String ROUTE = "/eth/v1/beacon/headers";
+  private final ChainDataProvider chainDataProvider;
+
+  public GetBlockHeaders(final DataProvider dataProvider, final JsonProvider jsonProvider) {
+    this(dataProvider.getChainDataProvider(), jsonProvider);
+  }
+
+  public GetBlockHeaders(
+      final ChainDataProvider chainDataProvider, final JsonProvider jsonProvider) {
+    super(jsonProvider);
+    this.chainDataProvider = chainDataProvider;
+  }
+
+  @OpenApi(
+      path = ROUTE,
+      method = HttpMethod.GET,
+      summary = "Get block headers",
+      tags = {TAG_V1_BEACON},
+      description =
+          "Retrieves block headers matching given query. By default it will fetch current head slot blocks.",
+      queryParams = {
+        @OpenApiParam(name = SLOT),
+        @OpenApiParam(name = PARENT_ROOT, description = "Not currently supported.")
+      },
+      responses = {
+        @OpenApiResponse(
+            status = RES_OK,
+            content = @OpenApiContent(from = GetBlockHeadersResponse.class)),
+        @OpenApiResponse(status = RES_BAD_REQUEST),
+        @OpenApiResponse(status = RES_INTERNAL_ERROR)
+      })
+  @Override
+  public void handle(@NotNull final Context ctx) throws Exception {
+    final Map<String, List<String>> queryParameters = ctx.queryParamMap();
+
+    final Optional<Bytes32> parentRoot =
+        SingleQueryParameterUtils.getParameterValueAsBytes32IfPresent(queryParameters, PARENT_ROOT);
+    final Optional<UInt64> slot =
+        SingleQueryParameterUtils.getParameterValueAsUInt64IfPresent(queryParameters, SLOT);
+
+    if (parentRoot.isPresent() && slot.isPresent()) {
+      ctx.status(SC_BAD_REQUEST);
+      ctx.result(
+          BadRequest.badRequest(
+              jsonProvider, "Either supply parent_root, slot, or neither, but not both."));
+      return;
+    }
```

Comment:
While not particularly useful, it's probably more typical to accept both params and treat it as an AND condition - so return only blocks at the specified slot with the specified parent root.

Reply:
the spec was super poorly defined... though we could do that, i'd probably just search on slot given our data structures?

Reply:
Yeah, search on slot then filter to check it matches the parentRoot supplied.

Reply:
Interesting question if the API is supposed to always default a head slot actually.  Maybe if you specify just `parentRoot` it should return blocks for the head slot that have the specified parent root which would be an awful lot easier to implement.  Not sure it means that but worth logging and issue to get it clarified and the docs updated.

Reply:
raised on the API https://github.com/ethereum/eth2.0-APIs/issues/103